File: Week_04/jump_game_ii/Solution.java
package jump_game_ii;

/**
 * This is an implicit bfs solution.
 * i == endpoint means you visited all the items on the current level.
 * Incrementing jumps++ is like incrementing the level you are on.
 * And endpoint = maxDist is like getting the queue size (level size) for the next level you are traversing.
 */
class Solution {
    public int jump(int[] nums) {
        int jumps = 0, endpoint = 0, maxDist = 0;
        for (int i = 0; i < nums.length - 1; i++) {
            maxDist = Math.max(maxDist, i + nums[i]);
            if (i == endpoint) {
                if (maxDist == endpoint) return -1;//说明不能跳到最后格子
                jumps++;
                endpoint = maxDist;
            }
        }
        return jumps;
    }
}

File: Week_04/jump_game_ii/Solution.py
from typing import List

# 贪心法
class Solution1:
    def jump(self, nums: List[int]) -> int:
        jmp, cur_end, max_loc = 0, 0, 0
        for loc in range(len(nums) - 1):
            max_loc = max(max_loc, loc + nums[loc])
            if loc == cur_end:
                jmp += 1
                cur_end = max_loc
        return jmp


# 贪心法，优化代码：
# 1.添加跳不到最后的情况，返回 -1
# 2.遇到能跳到最后的情况 max_loc >= endpoint，直接返回，不用继续循环到 endpoint
class Solution2:
    def jump(self, nums: List[int]) -> int:
        last = len(nums) - 1
        if last == 0: return 0

        jmp, cur_end = 0, 0
        loc, max_loc = 0, 0
        while loc <= max_loc:
            max_loc = max(max_loc, loc + nums[loc])
            if max_loc >= last:
                return jmp + 1
            if loc == cur_end:
                jmp += 1
                cur_end = max_loc
            loc += 1
        return -1
